pkg/deploy/gateway.go

```diff
@@ -0,0 +1,494 @@
+			logrus.Infof("Updating existing object: %s, name: %s", kind, actualMeta.GetName())
+			fmt.Printf("Difference:\n%s", diff)
+
+			err := clusterAPI.Client.Delete(context.TODO(), actual)
+			if err != nil {
+				return err
+			}
+
+			err = controllerutil.SetControllerReference(instance, blueprint, clusterAPI.Scheme)
+			if err != nil {
+				return err
+			}
+
+			err = clusterAPI.Client.Create(context.TODO(), blueprintObject)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func delete(clusterAPI ClusterAPI, obj metav1.Object) error {
+	key := client.ObjectKey{Name: obj.GetName(), Namespace: obj.GetNamespace()}
+	ro := obj.(runtime.Object)
+	if getErr := clusterAPI.Client.Get(context.TODO(), key, ro); getErr == nil {
+		if err := clusterAPI.Client.Delete(context.TODO(), ro); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetGatewayRouteConfig creates a config map with traefik configuration for a single new route.
+// `serviceName` is an arbitrary name identifying the configuration. This should be unique within operator. Che server only creates
+// new configuration for workspaces, so the name should not resemble any of the names created by the Che server.
+func GetGatewayRouteConfig(instance *orgv1.CheCluster, serviceName string, pathPrefix string, priority int, internalUrl string, stripPrefix bool) corev1.ConfigMap {
+	pathRewrite := pathPrefix != "/" && stripPrefix
+
+	data := `---
+http:
+  routers:
+    ` + serviceName + `:
+      rule: "PathPrefix(` + "`" + pathPrefix + "`" + `)"
+      service: ` + serviceName + `
+      priority: ` + strconv.Itoa(priority)
+
+	if pathRewrite {
+		data += `
+      middlewares:
+      - "` + serviceName + `"`
+	}
+
+	data += `
+  services:
+    ` + serviceName + `:
+      loadBalancer:
+        servers:
+        - url: '` + internalUrl + `'`
+
+	if pathRewrite {
+		data += `
+  middlewares:
+    ` + serviceName + `:
+      stripPrefix:
+        prefixes:
+        - "` + pathPrefix + `"`
+	}
+
+	return corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: instance.Namespace,
+			Labels: util.MergeMaps(
+				GetLabels(instance, gatewayConfigComponentName),
+				util.GetMapValue(instance.Spec.Server.SingleHostGatewayConfigMapLabels, DefaultSingleHostGatewayConfigMapLabels)),
+		},
+		Data: map[string]string{
+			serviceName + ".yml": data,
+		},
+	}
+}
+
+func DeleteGatewayRouteConfig(serviceName string, deployContext *DeployContext) error {
+	obj := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: deployContext.CheCluster.Namespace,
+		},
+	}
+
+	return delete(deployContext.ClusterAPI, obj)
+}
+
+// below functions declare the desired states of the various objects required for the gateway
+
+func getGatewayServerConfigSpec(instance *orgv1.CheCluster) corev1.ConfigMap {
+	return GetGatewayRouteConfig(instance, gatewayServerConfigName, "/", 1, "http://che-host:8080", false)
```

Comment:
is `che-host:8080` hardcoded when the service is created? Would it make sense to put it into variable so the actual service is synced with route in the code?